<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<HTML>
<HEAD>
<META name="bioperl project - developers & advocates of public domain tools for computational biolog, bioinformatics and life science research.">

<META name="keywords" content="perl, biology, bioinformatics, bioperl, computational biology, perl modules, object-oriented programming, genome, sequence analysis">

<TITLE>bio.perl.org - Script Central</TITLE>

</HEAD>
<BODY BGCOLOR="#FFFFFF" TEXT="#000000" LINK="#000080" VLINK="#000080" ALINK="#000080">

<font face="Arial, Helvetica, sans-serif" color="#000080">

<center>
<h2>bio.perl.org
</h2>
</center>

<TABLE BORDER=0>
<TR>
<TD BGCOLOR="#dadada" VALIGN=TOP WIDTH=100 ROWSPAN=20>
<!########################## SIDEBAR CONTENTS>
<font face="Arial, Helvetica, sans-serif" size=-1>
<a href="/">Home</a>
<p>
<a href="ftp://bio.perl.org/pub/">FTP Site</a>
<p>
<a href="/News/">News</a>
<p>
<a href="MailList.html">Mail Lists</a>
<p>
<a href="/UserInfo/">Developers</a>
<p>
<a href="/bioperl-bugs/">Bug Tracking</a>
<p>
<a href="/Participants/">Participants</a>
<p>
<a href="Thanks.html">Sponsors</a>
<p>
</font>
<!########################## SIDEBAR CONTENTS>
</TD>


<! margin >
<TD ROWSPAN=20><IMG SRC="/images/transparent.gif" WIDTH=20 ALT=" ">
</TD>


<TD ALIGN=LEFT VALIGN=TOP>
<DIV ALIGN=RIGHT></DIV>


<! main column #1>
<TABLE BORDER=0 CELLPADDING=3>

<TR>
<TD ALIGN=LEFT BGCOLOR="#dadada">
<B>
<font face="Arial, Helvetica, sans-serif" color="#000080">Freely available perl code for the life sciences {under construction}</font></B>
<BR>
</TD>
</TR>

<tr>
<td>&nbsp;</td>
</tr>

<tr>
<td ALIGN=LEFT BGCOLOR="#dadada">
<font face="Arial, Helvetica, sans-serif" color="#000080">
<b>A call for help:</b> Creating and keeping this index organized and current is a time consuming task. Help us out by letting us know if you spot a dead link, errors or know of some new software that has been released. Anyone wishing to help maintain or work on this list shoud contact <a href="mailto:dag@sonsorol.org">Chris Dagdigian</a>. 
</td>
</tr>

<tr>
<td>&nbsp;</td>
</tr>

<tr>
<td>
<b>Sanger Centre Modules</b><br>
<b>Author:</b> <a href="mailto:birney@sanger.ac.uk">Ewan Birney</a>, others.<br>
<b>URL:</b> <a href="http://www.sanger.ac.uk/Software/PerlModule/">http://www.sanger.ac.uk/Software/PerlModule/</a></br>
<b>Description:</b> Sequence and Alignment modules, EMBL modules, Feature Table & GFF modules, ACEdb modules.

<p>

<b>Will Fischer's Scripts</b><br>
<b>Author:</b> <a href="mailto:wfischer@indiana.edu">Will Fischer</a><br>
<b>URL:</b> <a href="http://hyrax.bio.indiana.edu:81/~wfischer/Perl_Scripts/">http://hyrax.bio.indiana.edu:81/~wfischer/Perl_Scripts/</a></br>
<b>Description:</b></br>
Standalone perl scripts for searching GenBank files by pattern,  translating nucleotide sequences to amino-acids (DNA to protein), making DNA alignments from amino-acid alignments, converting NEXUS files (from PAUP or MacClade) into FASTA or MEGA format, extracting the translations supplied by GenBank from CDS features in Genbank file
<p>

<b>Lincoln Stein's Modules</b><br>
<b>Author:</b><a href="mailto:lstein@cshl.org">Lincoln Stein</a><br>
<b>URL:</b>
<a href="http://stein.cshl.org/">http://stein.cshl.org/</a><br>
<a href="http://stein.cshl.org/software/boulder/">http://stein.cshl.org/software/boulder/</a><br>
<a href="http://stein.cshl.org/AcePerl/">http://stein.cshl.org/AcePerl/</a><br>
<a href="http://stein.cshl.org/software/boulder/">http://stein.cshl.org/software/boulder/</a><br>
<a href="http://stein.cshl.org/WWW/software/">http://stein.cshl.org/WWW/software/</a><br>

<b>Description:</b> Author of the famous <b>CGI::</b> modules also has code for the Boulder I/O format and perl interfaces to the ACEdb database.

<p>

<b>Jong Park's Perl Subroutine Library - Bio.pl</b><br>
<b>Author:</b><a href="mailto:jong@salt2.med.harvard.edu">Jong Park</a>, others.<br>
<b>URL:</b> <a href="http://cyrah.med.harvard.edu/~jong/Bio.pl.html">http://cyrah.med.harvard.edu/~jong/Bio.pl.html</a><br>
<b>Description:</b> A huge library of bio-related perl subroutines. A great place to find bits of useful code to add into your own programs.

<p>

<b>Ian Korf's Code & Bio::GSC Modules</b><br>
<b>Author:</b> <a href="mailto:ikorf@sapiens.wustl.eduikorf">Ian Korf</a><br>
<b>URL:</b> <a href="http://sapiens.wustl.edu/~ikorf/bioinformatics.html">http://sapiens.wustl.edu/~ikorf/bioinformatics.html</a>
<b>Description:</b> Ian has made public the Bio::GSC modules that the WashU Genome Sequencing Center uses. He also has a standalone Blast parser called BP2.pl.<br>

<p>

<b>Roland Walker's SEALS system</b><br>
<b>Author:</b> <a href="mailto:walker@ncbi.nlm.nih.gov?subject=SEALS">Roland Walker</a><br>
<b>URL:</b> <a href="http://www.ncbi.nlm.nih.gov/Walker/SEALS/">http://www.ncbi.nlm.nih.gov/Walker/SEALS/</a><br>
<b>Description:</b> SEALS (A System for Easy Analysis of Lots of Sequences) is a perl software package expressly designed for large-scale research projects in bioinformatics.

<p>

<b>Paintblast.pm</b><br>
<b>Author:</b> <a href="mailto:guffanti@tigem.it">Alessandro Guffanti</a><br>
<b>URL:</b> <a href="http://hercules.tigem.it/Biomodules.html">http://hercules.tigem.it/Biomodules.html</a><br>
<b>Description:</b> Parses a Blast output and produces a GIF image with a schematic representation of the resulting alignments. Module is still experimental.<br>
<font color="red"><b>REQUIRES:</b></font> needs <A HREF="http://stein.cshl.org/WWW/software/GD/GD.html">Lincoln
Stein's GD module.</A> GD module requires Tom Boutell's graphic <A HREF="http://www.boutell.com/gd/gd.html">libgd library</A>.<br>

<p>

<b>CheckSeq.pm</b><br>
<b>Author:</b> <a href="mailto:guffanti@tigem.it">Alessandro Guffanti</a><br>
<b>URL:</b> <a href="http://hercules.tigem.it/Biomodules.html">http://hercules.tigem.it/Biomodules.html</a><br>
<b>Description:</b> Module examines an input sequence, reformats it in FastA format and determines whether it is a DNA or Protein sequence. It then checks the coherency between the sequence type and the chosen blast program (blastn, blastp, tblastn, blastx, tblastx). The module was created in order to avoid potential problems for users of web-based Blast-based services. <br>
<font color="red"><b>REQUIRES:</b></font> module uses the fmtseq program from <A HREF="ftp://ftp.cs.ucdavis.edu/pub/strings/seqio.tar.gz">J.R.Knight's SeqIo package.

<p>

</td>
</tr>


<TR>
<TD>
<p>
<font face="Arial, Helvetica, sans-serif" size=+1>
<a href="/">&lt;&lt;back</a>
</font>
<p>
</font>
</TD>
</TR>



<!spacer >
<TR><TD><IMG SRC="/images/transparent.gif" WIDTH=230 HEIGHT=1></TD></TR>
</TABLE>
</TD>

<! end first main column>


<! vertical spacer bar>
<TD WIDTH=3 BGCOLOR="#000080"><IMG SRC="/images/transparent.gif" WIDTH=1 HEIGHT=1></TD>


</TD>
</TR>
</TABLE>

<! ### START CVS DATESTAMP >
<p><p align=right>Last Modified: $Date$</p>
<! ### STOP CVS DATESTAMP >

</BODY>
</HTML>




