<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<html>
  <head>
<META name="bioperl project - developers & advocates of public domain tools for computational biology, bioinformatics and life science research.">

<META name="keywords" content="perl, biology, bioinformatics, bioperl, computational biology, perl modules, object-oriented programming, genome, sequence analysis">

    <title>bio.perl.org - Script Central</title>
  </head>

  <BODY BGCOLOR="#FFFFFF" TEXT="#000000" LINK="#000080" VLINK="#000080" ALINK="#000080">
  


<font face="Arial, Helvetica, sans-serif" color="#000080">

<center>
<h2>
<a href="http://bio.perl.org/">bio.perl.org</a>
</h2>
</center>

<TABLE BORDER=0>
<TR>
<TD BGCOLOR="#dadada" VALIGN=TOP WIDTH=100 ROWSPAN=20>
<!########################## SIDEBAR CONTENTS>
<font face="Arial, Helvetica, sans-serif">
<a href="/">Home</a>
<p>
<a href="/Projects/">Projects</a>
<p>
<a href="/Projects/modules.html">Docs</a>
<p>
<a href="http://bio.perl.org/Core/Latest/">Download</a>
<p>
<a href="/News/">News</a>
<p>
<a href="/MailList.html">Mail Lists</a>
<p>
<a href="/UserInfo/">Developers</a>
<p>
<a href="/Bugs/">Bugs</a>
<p>
<a href="/Participants/">Participants</a>
<p>
<a href="/Related.html">Related</a>
<p>
<a href="/Thanks.html">Sponsors</a>
<p>
</font>
<!########################## SIDEBAR CONTENTS>
</TD>


<! margin >
<TD ROWSPAN=20><IMG SRC="/images/transparent.gif" WIDTH=5 ALT=" ">
</TD>


<TD ALIGN=LEFT VALIGN=TOP>
<DIV ALIGN=RIGHT></DIV>


<! main column #1>
<TABLE BORDER=0 CELLPADDING=3>

<TR>
<TD ALIGN=LEFT BGCOLOR="#dadada">
<B>
<font face="Arial, Helvetica, sans-serif" color="#000080">ScriptCentral - Freely available perl code for the life sciences</font></B>
<BR>
</TD>
</TR>

<tr>
<td>&nbsp;</td>
</tr>

<tr>
<td ALIGN=LEFT BGCOLOR="#dadada">
<font face="Arial, Helvetica, sans-serif" color="#000080">
<b>A call for help:</b> Creating and keeping this index organized and current is a time consuming task. Help us out by letting us know if you spot a dead link, errors or know of some new software that has been released. If you have scripts that you would like to share, please feel free to contact us as well --  <a href="mailto:alm13@cornell.edu">Andreas Matern</a>. 
</td>
</tr>

<tr>
<td>&nbsp;</td>
</tr>

<tr>
<td>


<p>
<b>Will Fischer's Scripts</b><br>
<b>Author:</b> <a href="mailto:wfischer@indiana.edu">Will Fischer</a><br>
<b>URL:</b> <a href="http://sunflower.bio.indiana.edu/~wfischer/Perl_Scripts/">http://sunflower.bio.indiana.edu/~wfischer/Perl_Scripts/</a></br>
<b>Description:</b></br>
Standalone perl scripts for searching GenBank files by pattern,  translating nucleotide sequences to amino-acids (DNA to protein), making DNA alignments from amino-acid alignments, converting NEXUS files (from PAUP or MacClade) into FASTA or MEGA format, extracting the translations supplied by GenBank from CDS features in Genbank file
<p>

<b>Lincoln Stein's Modules</b><br>
<b>Author:</b><a href="mailto:lstein@cshl.org">Lincoln Stein</a><br>
<b>URL:</b>
<a href="http://stein.cshl.org/">http://stein.cshl.org/</a><br>
<a href="http://stein.cshl.org/software/boulder/">http://stein.cshl.org/software/boulder/</a><br>
<a href="http://stein.cshl.org/AcePerl/">http://stein.cshl.org/AcePerl/</a><br>

<a href="http://stein.cshl.org/WWW/software/">http://stein.cshl.org/WWW/software/</a><br>

<b>Description:</b> Author of the famous <b>CGI::</b> modules also has code for the Boulder I/O format and perl interfaces to the ACEdb database.

<p>

<b>Jong Park's Perl Subroutine Library - Bio.pl</b><br>
<b>Author:</b><a href="mailto:jong@ebi.ac.uk">Jong Park</a>, others.<br>
<b>URL:</b> <a href="http://www.mrc-lmb.cam.ac.uk/genomes/jong/Bio.pl.html">http://www.mrc-lmb.cam.ac.uk/genomes/jong/Bio.pl.html</a><br>
<b>Description:</b> A huge library of bio-related perl subroutines. A great place to find bits of useful code to add into your own programs.


<p>

<b>Ian Korf's Code & Bio::GSC Modules</b><br>
<b>Author:</b> <a href="mailto:ikorf@sapiens.wustl.eduikorf">Ian Korf</a><br>
<b>URL:</b> <a href="http://sapiens.wustl.edu/~ikorf/bioinformatics.html">http://sapiens.wustl.edu/~ikorf/bioinformatics.html</a><br>
<b>Description:</b> Ian has made public the Bio::GSC modules that the WashU Genome Sequencing Center uses. He also has a standalone Blast parser called BP2.pl.<br>

<p>

<b>Roland Walker's SEALS system</b><br>
<b>Author:</b> <a href="mailto:walker@ncbi.nlm.nih.gov?subject=SEALS">Roland Walker</a><br>
<b>URL:</b> <a href="http://www.ncbi.nlm.nih.gov/Walker/SEALS/">http://www.ncbi.nlm.nih.gov/Walker/SEALS/</a><br>
<b>Description:</b> SEALS (A System for Easy Analysis of Lots of Sequences) is a perl software package expressly designed for large-scale research projects in bioinformatics.

<p>

<b>Paintblast.pm</b><br>
<b>Author:</b> <a href="mailto:guffanti@tigem.it">Alessandro Guffanti</a><br>
<b>URL:</b> <a href="http://hercules.tigem.it/Biomodules.html">http://hercules.tigem.it/Biomodules.html</a><br>
<b>Description:</b> Parses a Blast output and produces a GIF image with a schematic representation of the resulting alignments. Module is still experimental.<br>
<font color="red"><b>REQUIRES:</b></font> needs <A HREF="http://stein.cshl.org/WWW/software/GD/GD.html">Lincoln
Stein's GD module.</A> GD module requires Tom Boutell's graphic <A HREF="http://www.boutell.com/gd/gd.html">libgd library</A>.<br>

<p>

<b>CheckSeq.pm</b><br>
<b>Author:</b> <a href="mailto:guffanti@tigem.it">Alessandro Guffanti</a><br>
<b>URL:</b> <a href="http://hercules.tigem.it/Biomodules.html">http://hercules.tigem.it/Biomodules.html</a><br>
<b>Description:</b> Module examines an input sequence, reformats it in FastA format and determines whether it is a DNA or Protein sequence. It then checks the coherency between the sequence type and the chosen blast program (blastn, blastp, tblastn, blastx, tblastx). The module was created in order to avoid potential problems for users of web-based Blast-based services. <br>
<font color="red"><b>REQUIRES:</b></font> module uses the fmtseq program from <A HREF="http://bioinformatics.weizmann.ac.il/software/strings/">J.R.Knight's SeqIo package.</A><br>

<p>
<b>TBOB</b><br>
<b>Author:</b> <a href="mailto:riouxp@bchumontreal.ca">Pierre Rioux</a></br>
<b>URL:</b> <a href="ftp://megasun.bch.umontreal.ca/pub/tbob">ftp://megasun.bch.umontreal.ca/pub/tbob</A><br>
<b>Description:</b> Text based BLAST output Browser.  Output from the BLAST programs can be difficult to read, especially when there are many search results  to be analyzed, each potentially with many matches with sequences in the database. TBOB solves this problem by only showing parts of the BLAST output at any time, letting the user jump from one place to another using simple one-keystoke commands.  TBOB also permits retrieval of sequences from the NCBI retrieve e-mail server, so that interesting sequences can be obtained for further analysis. <br>
<P>

<P>
<b>MOLPHY</B><br>
<b>Author:</b> <a href="mailto:adachi@ism.ac.jp">Jun Adachi</A><br>
<b>URL:</b> <a href="ftp://sunmh.ism.ac.jp/pub/molphy">ftp://sunmh.ism.ac.jp/pub/molphy</A><br>
<b>Description:</b> MOLPHY is a program package for MOLecular PHYlogenetics.  PROTML is a main program in MOLPHY for inferring evolutionary trees from PROTein sequences by using the Maximum Likelihood method.  The MOLPHY package includes a number of useful Perl Utilites for DNA/Protein sequence manipulation.<br>

<P>
<b>PARBOOT</b><br>
<b>Author:</b>  <a href="mailto:riouxp@bchumontreal.ca">Pierre Rioux</a></br>
<b>URL:</b> <a href="ftp://megasun.bch.umontreal.ca/pub/parboot">ftp://megasun.bch.umontreal.ca/pub/parboot</A><br>
<b>Description:</b> PARBOOT is a parallel implementaion of the bootstrapping function available in many of the PHYLIP phylogenetic analysis programs. The PARBOOT application splits a multiple-dataset input file into its independent datasets, and processes these datasets in parallel on multiple hosts (or one host with multiple CPUs). This means that the speed of a bootsptrap analysis can be approximately improved in proportion to the number and power of CPUs available to the user.<br> 
<font color="red"><b>REQUIRES:</b></font> rsh, rcp, <A href="http://evolution.genetics.washington.edu/phylip.html">PHYLIP</A><br>

<p>
<b>SLBC - Search Launcher Batch Client</b><br>
<b>Author:</b> <a href="mailto:rsmith@bcm.tmc.edu">Randall F. Smith</a><br>
<b>URL:</b> <a href="http://dot.imgen.bcm.tmc.edu:9331/downloads/software/batch_client.html">http://dot.imgen.bcm.tmc.edu:9331/downloads/software/batch_client.html</A><br>
<b>Description:</b> The SLBC will read in multiple input files, perform a specified search (e.g., BLAST, BEAUTY) on each sequence (on at a time) then store the result files on the user's system as individual HTML documents. The search results can then be viewed using an HTML viewer.  The search results will be the same as if the search was performed from the <A href="http://www.hgsc.bcm.tmc.edu/SearchLauncher/">Baylor College of Medicine's Search Launcher Web Pages.</A><br>

<P>
<B>GEANFAMMER</b> - GEnome ANalysis and Protein FAMily MakER<br>
<b>Author:</b> <A HREF="mailto:jong@ebi.ac.uk">Jong Park</a><br>
<b>URL:</b> <A href="http://www.mrc-lmb.cam.ac.uk/genomes/geanfammer.html">http://www.mrc-lmb.cam.ac.uk/genomes/geanfammer.html</A><br>
<b>Description:</b> Geanfammer has been developed for the analysis of most of the complete bacterial genomes announced since 1995. It summarizes the whole procedure of preparing statistically and biologically more relevant protein (sequence) duplication modules before any more biological analysis like structure and fucntion assignment.<br>

<P>





</td>
</tr>


<TR>
<TD>
<p>
<font face="Arial, Helvetica, sans-serif" size=+1>
<a href="/">&lt;&lt;back</a>
</font>
<p>
</font>
</TD>
</TR>



<!spacer >
<TR><TD><IMG SRC="/images/transparent.gif" WIDTH=230 HEIGHT=1></TD></TR>
</TABLE>
</TD>

<! end first main column>


<! vertical spacer bar>
<TD WIDTH=3 BGCOLOR="#000080"><IMG SRC="/images/transparent.gif" WIDTH=1 HEIGHT=1></TD>


</TD>
</TR>
</TABLE>

<! ### START CVS DATESTAMP >
<p><p align=right>Last Modified: $Date$</p>
<! ### STOP CVS DATESTAMP >


    <hr>
    <address><a href="mailto:alm13@cornell.edu">Andreas Matern</a></address>
<!-- Created: Wed Sep  8 14:19:35 EDT 1999 -->
<!-- hhmts start -->
Last modified: Thu Sep  9 14:45:02 EDT 1999
<!-- hhmts end -->
  </body>
</html>
