<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<HTML>
<HEAD>
<META name="bioperl project - developers & advocates of public domain tools for computational biolog, bioinformatics and life science research.">

<META name="keywords" content="perl, biology, bioinformatics, bioperl, computational biology, perl modules, object-oriented programming, genome, sequence analysis">

<TITLE>bio.perl.org - Structure Requirements</TITLE>

</HEAD>
<BODY BGCOLOR="#FFFFFF" TEXT="#000000" LINK="#000080" VLINK="#000080" ALINK="#000080">

<font face="Arial, Helvetica, sans-serif" color="#000000">

<center>
<h2><font color="#000080">bio.perl.org </font>
</h2>

<h2>
Preliminary requirements list for a Structure object
</h2>

</center>

<hr>

This is an initial listing. It is not necessarily complete. Not all of
the things listed here are expected to go into the Structure object
itself. Think of this is a rough set of use-cases for biomolecular
structure information, which should be considered when designing a
structure object.

<p>
<b>Send suggestions to <a href="mailto:sac@neomorphic.com">sac@neomorphic.com</a>

<p>
The requirements seem to fit into 5 main categories:
<ol type="I">
<li><a href="#atoms">Atomic coordinates</a>
<li><a href="#features">Structural features</a>
<li><a href="#anal">Analyzing & editing structures</a>
<li><a href="#seq">Sequence information</a>
<li><a href="#db">Database information</a>
</ol>

<!--     Atomic Coordinates Requirements     -->

<p>
<ol type="I">
<h3>
<a name="atoms">
<li>Atomic Coordinates
</a>
</h3>

<ol type="1">

<li>Parse and output different structure file formats, at a
  minimum PDB but could include mmCIF, CML/XML, and other formats.
<li>Be able to obtain different representations the structural coordinates.
  For example, some applications require a hierarchical representation (complex, 
  molecule, domain, residue, atom) while other might require a
  non-hierarchically representation (a simple array of atoms or a
  graph of atoms and bonds). Some applications may utilize minimal representations of the 
   structure, such as only backbone atoms (C-alpha or mainchain form
proteins; phosphodiester backbone for nucleic acids) or heavy-atom
only (no hydrogens, no waters).
<li>Convert between alternate representations of the structure. 
<li>Molecular coverage: at a minimum, support protein and nucleic acid xtallograpic and NMR
  structures and their heterogens. Multi-chain and multi-model (NMR) structures
 should also be supported.
<li>Theoretical structural models: Obtain quality factors and other
relevant modeling statistics. this will depend on the modeling
method. For a homology-based model, methods should be available to get
the structure and region of the structure the model is based on.

</ol>

<!--     Structural Feature Requirements     -->

<h3>
<a name="features">
<li>Structural Features
</a>
</h3>

<ol type="1">

<li>Provide methods for querying structure for statistics regarding
  atoms, hetatoms, connectivity, secondary structural elements,
  disulfide bonds, presence of alternate conformations, structure
  quality factors (resolution, R factor), crystal packing group,
  and other data obtainable directly from a PDB file. 
  Specific  examples:
   <ul>
   <li>get the  start  and end residue numbers for each chain in a
      structure.
   <li>get the chain association for a given hetero group.
   </ul>
<li> Obtain all atomic details for a given atom (xyz, atom serial
  number, atom name, chain name, residue name, residue number, occupancy, B-factor).
  For example, given an atom, determine what residue and what molecule
  (chain) it is in.
<li> Obtain information regarding what domains and other structurally
  or biologically relevant information such as the locations of active sites,
  metal binding site, covalently modified residues, etc.
<li>Need a robust domain-definition mechanism:
 <ul>
<li> how defined (SCOP, CATH, etc.)
<li> interchain, intra-chain, contiguous, discontiguous.
<li> Many of the requirements of the structure object apply
    at the level of the individual domain as well.
</ul>
<li>Need a flexible mechanism for specifying locations of features.

</ol>

<!--     Analyzing & Editing Structures -->

<h3>
<a name="anal">
<li>Analyzing & Editing Structures
</a>
</h3>
Structural analyses can be rather compute intensive and so it makes
sense to not burden the structure object with complex analytical
methods. Such methods are best left for a structural utility module or
application. Still, it's useful to consider these analysis from the
point of view of what requirements they impose on a structure object.
<ol type="1">

<li>Display/select acidic, basic, polar and non-polar amino acids,
hydrogen  bonds
<li>Charge molecule and compute energy, hydrogen bonds, salt bridges.
<li>Given sequence alignment, compute optimal structure superposition
(produce a rotation/translation matrix and/or RMSD).
<li>Given a matrix, compute the RMSD.
<li>Add/remove/change residues, hydrogens, and water molecules.
<li>Energy minimize an altered structure.
</ol>


<!--     Sequence Information Requirements     -->

<h3>
<a name="seq">
<li>Sequence Information
</a>
</h3>

<ol type="1">

<li> Obtain a representation of the structure as sequence or
  a set of sequences (strings as well as specialized objects).
<li> Get sequence from the actual set of residues in the atomic
coordinate set.
<li> Map between residue numbers between the atomic coordinate set
and the numbers in the sequence string (say from a Fasta
representation). 
<li> Specify and generate subsections of the structure given a
  subsequence (range of integer positions as well as a sequence object).

</ol>



<!--     Database Information Requirements     -->

<h3>
<a name="db">
<li>Database Information
</a>
</h3>

<ol type="1">

<li> Build a structure object given a valid structural database identifier.
<li> Obtain information about the structure given a valid structural
  database identifier (molecule name, source organism, deposition
  date, etc.). 
<li> Support local and remote database access. Many groups store
versions of the PDB locally or may have their own private structural database.
<li> Get URLs for web-based information about the structure.


</ol>

</ol>

<hr>

<! ### START CVS DATESTAMP >
<p><p align=right>Last Modified: $Date$</p>
<! ### STOP CVS DATESTAMP >

<! ### START CVS REVISION >
<!-- $Id$ -->
<! ### STOP CVS REVISION >


</body>
</html>

