<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<HTML>
<HEAD>
<META name="bioperl project - developers & advocates of public domain tools for computational biolog, bioinformatics and life science research.">

<META name="keywords" content="perl, biology, bioinformatics, bioperl,
computational biology, perl modules, object-oriented programming, 3D
structure, macromolecule, requirements">

<TITLE>bio.perl.org - Structure Requirements</TITLE>

</HEAD>
<BODY BGCOLOR="#FFFFFF" TEXT="#000000" LINK="#000080" VLINK="#000080" ALINK="#000080">

<font face="Arial, Helvetica, sans-serif" color="#000000">

<center>
<h2>
<font color="#000080">
<a href="http://bio.perl.org/">bio.perl.org</a>
</font>
</h2>

<h2>
Preliminary Requirements for a Structure Object
</h2>

<a href="http://bio.perl.org">Bioperl</a> |
<a href="http://bio.perl.org/Projects/">Projects</a> |
<a href="http://bio.perl.org/Projects/Structure">Structure</a> 
</center>

<hr>

This is an initial listing of requirements for a software component or
set of components to represent biological macromolecular 3D
structural information.  It is not necessarily complete. Think of
this is a rough set of use-cases for biomolecular structure
information, which should be considered when designing a structure
object.
<p>
One of the purposes of this requirements listing is to define what
should be part of the structure object and what should not.  Not all
of the functionalities listed here are expected to go into the
structure object itself but instead could be provided by external
utilities. The structure object could provide an interface to such
utilities and delegate tasks to perform the actual work as
necessary.
<p>
Another important issue is whether to have lightweight and heavyweight
versions of a structure object, as is being done for 
<a href="http://bio.perl.org/Projects/Sequence">sequence objects</a> and
<a href="http://bio.perl.org/Projects/SeqAlign">sequence alignment objects</a>.
How would you distinguish light- and heavyweight versions of a structure object?

<p>
<b>Send suggestions to <a href="mailto:vsns-bcd-perl@lists.uni-bielefeld.de">vsns-bcd-perl@lists.uni-bielefeld.de</a>
<p>
View responses (starting 12/98) at
<a href="http://www.uni-bielefeld.de/mailinglists/BCD/vsns-bcd-perl/">http://www.uni-bielefeld.de/mailinglists/BCD/vsns-bcd-perl/</a>
</b>

<p>
The requirements seem to fit into 5 main categories:
<ol type="I">
<li><a href="#atoms">Atomic coordinates</a>
<li><a href="#features">Structural features</a>
<li><a href="#anal">Analyzing & editing structures</a>
<li><a href="#seq">Sequence information</a>
<li><a href="#db">Database-related</a>
</ol>

<!--     Atomic Coordinates Requirements     -->

<p>
<ol type="I">
<h3>
<a name="atoms">
<li>Atomic Coordinates
</a>
</h3>

<ol type="1">

<p>
<li>Parse and output different structure file formats, at a
  minimum PDB but could include mmCIF, CML/XML, and other formats.
<p>
<li>Be able to obtain different representations the structural coordinates.
  For example, some applications require a hierarchical representation (complex, 
  molecule, domain, residue, atom) while other might require a
  non-hierarchical representation (a simple array of atoms or a
  graph of atoms and bonds). Some applications may utilize minimal representations of the 
   structure, such as only backbone atoms (C-alpha or mainchain for
proteins; phosphodiester backbone for nucleic acids) or heavy-atom
only (no hydrogens, no waters) or a set of vectors representing
  secondary structural elements.
<p>
<li>Convert between alternate representations of the structure. 
<p>
<li>Molecular coverage: at a minimum, support protein and nucleic acid crystallograpic and NMR
  structures and their heterogens. Multi-chain and multi-model (NMR) structures
 should also be supported. The focus here is on <b>macro</b>molecular
  structure, but small molecule structures should be supported to the
  extent that they interact with biological macromolecules.
<p>
<li>Theoretical structural models: Obtain quality factors and other
relevant modeling statistics. This will depend on the modeling
method. For a homology-based model, methods should be available to get
the structure and region of the structure on which the model is based.

</ol>

<!--     Structural Feature Requirements     -->

<h3>
<a name="features">
<li>Structural Features
</a>
</h3>

<ol type="1">

<li>Provide methods for querying structure for statistics regarding
  atoms, hetatoms, connectivity, secondary structural elements,
  disulfide bonds, presence of alternate conformations, structure
  quality factors (resolution, R factor), crystal packing group,
  and other data obtainable directly from a PDB file. 
<p>
  Specific  examples:
   <ul>
   <li>Get the  start  and end residue numbers for each chain in a
      structure.
   <li>Get the chain association for a given hetero group.
   </ul>
<p>
<li> Obtain all atomic details for a given atom (xyz, atom serial
  number, atom name, chain name, residue name, residue number, occupancy, B-factor).
  For example, given an atom, determine what residue and what molecule
  (chain) it is in.
<p>
<li>Obtain atoms or residues based on physical-chemical
properties: acidic, basic, polar, non-polar amino acids, spcific types
of residues or atoms (e.g., oxygen, glycine),  hydrogen
bonds, disulfide bonds, etc.
<p>
<li> Obtain information regarding what domains and other structurally
  or biologically relevant information such as the locations of active sites,
  metal binding site, covalently modified residues, etc.
<p>
<li>Need a robust domain-definition mechanism:
 <ul>
<li> how defined (SCOP, CATH, etc.)
<li> interchain, intra-chain, contiguous, discontiguous.
<li> Many of the requirements of the structure object apply
    at the level of the individual domain as well (e.g., coordinates
    representations, analysis methods, features, etc.).
</ul>
<p>
<li>Need a flexible mechanism for specifying locations of features.

</ol>

<!--     Analyzing & Editing Structures -->

<h3>
<a name="anal">
<li>Analyzing & Editing Structures
</a>
</h3>
Structural analyses can be rather compute intensive and so it makes
sense to not burden the structure object with complex analytical
methods. Such methods are best left for a structural utility module or
application. Still, it's useful to consider these analyses from the
point of view of what requirements they impose on a structure object.
<p>
<ol type="1">
<li>Calculate dihedral angles (for Ramachandran plot, for example).
<p>
<li>Get a list of atoms or residues within a given radius from a
specified atom or residue.
<p>
<li>Charge molecule and compute energy, hydrogen bonds, salt bridges.
<p>
<li>Spatial calculations: solvent accessible surface, volume, cavities,
etc. 
<p>
<li>Determine the axes of secondary structural elements.
<p>
<li>Given sequence alignment, compute optimal structure superposition
(produce a rotation/translation matrix and/or RMSD).
<p>
<li>Given a matrix, compute the RMSD.
<p>
<li>Add/remove/change residues, hydrogens, and water molecules.
<p>
<li>Energy minimize an altered structure.
<p>
<li>Dock structures (identify interacting surfaces between
two given structures or screen a library of small molecules against a
macromolecule, perhaps constraining the interacting portion of the
macromolecule).
</ol>


<!--     Sequence Information Requirements     -->

<h3>
<a name="seq">
<li>Sequence Information
</a>
</h3>

<ol type="1">

<li> Obtain a representation of the structure as sequence or
  a set of sequences (strings as well as specialized objects).
<p>
<li> Get sequence from the actual set of residues in the atomic
coordinate set.
<p>
<li> Map between residue numbers in the atomic coordinate set
and the numbers in the sequence string (say from a Fasta
representation). 
<p>
<li> Specify and generate subsections of the structure given a
  subsequence (range of integer positions as well as a sequence object).
<p>
<li>Obtain an arbitrary set of residues on a structure.

</ol>



<!--     Database-Related Requirements     -->

<h3>
<a name="db">
<li>Database-Related
</a>
</h3>

<ol type="1">

<li> Build a structure object given a valid structural database identifier.
<p>
<li> Obtain information about the structure given a valid structural
  database identifier (molecule name, source organism, deposition
  date, dbxref etc.). 
<p>
<li>Get related structures and domains based on criteria such as
  sequence similarity, structural similarity, date deposited,
  resolution, organism name, molecule name, etc.
<p>
<li> Get URLs for web-based information about the structure.
<p>
<li> Support local and remote database access. Many groups store
versions of the PDB locally or may have their own private structural database.


</ol>

</ol>

<hr>

<! ### START CVS DATESTAMP >
<p><p align=right>Last Modified: $Date$</p>
<! ### STOP CVS DATESTAMP >

<! ### START CVS REVISION >
<!-- $Id$ -->
<! ### STOP CVS REVISION >


</body>
</html>

