<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
<html>

<head>
<meta
name="bioperl project - developers &amp; advocates of public domain tools for computational biolog, bioinformatics and life science research.">
<meta name="keywords"
content="perl, biology, bioinformatics, bioperl, computational biology, perl modules, object-oriented programming, genome, sequence analysis">
<title>bio.perl.org - Projects</title>
</head>

<body BGCOLOR="#FFFFFF" TEXT="#000000" LINK="#000080" VLINK="#000080" ALINK="#000080">
<font face="Arial, Helvetica, sans-serif" color="#000080">

<h2 align="center"><a href="http://bio.perl.org/">bio.perl.org</a> </h2>

<table BORDER="0">
  <tr>
    <td BGCOLOR="#dadada" VALIGN="TOP" WIDTH="100" ROWSPAN="20"><!########################## SIDEBAR CONTENTS> <font
    face="Arial, Helvetica, sans-serif"><a href="/">Home</a> <p><b><a href="/Projects">Projects</a>
    </b></p>
    <p><a href="/Core/Latest/modules.html">Docs</a> </p>
    <p><a href="http://bio.perl.org/Core/Latest/">Download</a> </p>
    <p><a href="/News/">News</a> </p>
    <p><a href="/MailList.html">Mail Lists</a> </p>
    <p><a href="/UserInfo/">Developers</a> </p>
    <p><a href="/Bugs/">Bugs</a> </p>
    <p><a href="/Participants/">Participants</a> </p>
    <p><a href="/Related.html">Related</a> </p>
    <p><a href="/Thanks.html">Sponsors</a> </p>
    </font><p><!########################## SIDEBAR CONTENTS> </td>
<! margin >
    <td ROWSPAN="20"><img SRC="/images/transparent.gif" WIDTH="5" ALT=" "> </td>
    <td ALIGN="LEFT" VALIGN="TOP"><div ALIGN="RIGHT"></div><! main column #1><p></p>
    </font><table BORDER="0" CELLPADDING="3" width="857">
      <tr>
        <td BGCOLOR="#dadada" colspan="2" width="849"><font face="Arial, Helvetica, sans-serif"
        color="#000080"><b>Bioperl Projects: XML</b><br>
        </font></td>
      </tr>
      <tr>
        <td width="234" valign="top"><h2><font color="#0000A0">Contents</font></h2>
        <ul>
          <li><font color="#0000A0"><a href="#Purpose">Purpose</a></font></li>
          <li><font color="#0000A0"><a href="#FAQ">FAQ</a></font></li>
          <li><font color="#0000A0"><a href="#Software">Software</a></font></li>
          <li><font color="#0000A0"><a href="#discussion">Discussion groups</a></font></li>
          <li><font color="#0000A0"><a href="#Related">Related sites</a></font></li>
          <li><font color="#0000A0"><a href="#Site">About this site</a></font></li>
        </ul>
        </td>
        <td width="607" valign="top"><h2><font color="#0000A0"><a name="Purpose">Purpose</a></font></h2>
        <p><font color="#0000A0">The purpose of this web site is to coordinate information
        dissemination and discussion on the use of XML in bioinformatics, and to serve as a
        repository for metadata describing the relationships between&nbsp; biological XML data
        resources.&nbsp; My current thinking is that the latter should be accomplished through <a
        href="http://w3c.org/RDF">RDF</a>, but since this is still in draft stage, this thinking
        may change.</font></p>
        <h2><font color="#0000A0"><a name="FAQ">FAQ</a> (<a
        href="http://www.perlxml.com/faq/perl-xml-faq.html">perl XML FAQ</a>)</font></h2>
        <blockquote>
          <h4><font color="#0000A0">What problems can XML solve?</font></h4>
          <blockquote>
            <p><font color="#0000A0">Biology is a complex discipline and a wide variety of data
            resources and repositories have been developed to support biological research.&nbsp; Many
            of these are interrelated, but it is currently difficult to identify or use these in
            relationships computationally because the different data sources use incompatible formats
            and semantics.&nbsp; XML offers a way to&nbsp; serve and describe data in a uniform and
            automatically parseable format.</font></p>
            <p><font color="#0000A0">XML also provides an open source solution for data migration
            between programming languages.&nbsp; Using modules such ash <a href="#XML::Parser">XML::Parser</a>,
            it is possible to move a structured data object between C/C++, java, and PERL without
            having to worry about the encoding/decoding details.</font></p>
          </blockquote>
          <h4><font color="#0000A0">What problems will not be solved by XML?</font></h4>
          <blockquote>
            <p><font color="#0000A0">XML is basically a language for describing data, but it is not
            the description.&nbsp; We need to develop that ourselves.&nbsp; Just because <a
            href="#bsml">BSML</a> and <a href="#bioml">BIOML</a> both use XML to encode data does not
            mean that a <a href="#bsml">BSML</a> browser will be able to view <a href="#bioml">BIOML</a>
            data.&nbsp; The fact that both are using XML means that writing a filter to convert one to
            the other can skip alot of the low level parsing that would otherwise be needed.</font></p>
          </blockquote>
          <h4><font color="#0000A0">Why use XML in biocomputing?</font></h4>
          <blockquote>
            <p><font color="#0000A0">XML is an easily and automatically parseable way to present data
            on the web.&nbsp; The basic representation uses standard ASCII text so you can view, edit,
            and debug data with standard text tools like emacs, vi, grep, and wc. Most importantely, a
            strong industry consortium is developing that will support the use of XML as the
            underlying technology for a next generation internet.</font></p>
          </blockquote>
          <h4><font color="#0000A0">What is the relationship of XML to BIO.PERL?</font></h4>
          <blockquote>
            <p><font color="#0000A0">At the moment, BIO.PERL.org is &quot;looking with interest&quot;
            at XML, but there is no formal relationship.&nbsp; PERL is a good tool to use with XML
            (see modules below), and XML hopefully become a standard for biological data resources, so
            BIO.PERL may, in the future, incorporate XML technology into the BIO.PERL toolkit.</font></p>
          </blockquote>
          <h4><font color="#0000A0">What is the status of XML?</font></h4>
          <blockquote>
            <p><font color="#0000A0">XML itself is a well-defined standard that has been promulgated
            by the W3C organization.&nbsp; At present, few tools or data resources making use of this
            standard are available.&nbsp; One of the goals of the XML initiative is to facilitate the
            development of metadata resources (resources describing other resources), but at present
            this field is in rapid flux and the <a href="http://w3c.org/RDF">RDF</a> format itself
            remains in draft stage.</font></p>
          </blockquote>
          <h4><font color="#0000A0">Is XML supported by the major databases in biology?</font></h4>
          <blockquote>
            <p><font color="#0000A0">At present, none of the major database support an XML interface,
            although almost all of them are actively exploring the use of XML for future work.</font></p>
          </blockquote>
          <h4><font color="#0000A0">Correctness in XML - well formed &amp; valid</font></h4>
          <blockquote>
            <p><font color="#0000A0">XML data can have varying levels of &quot;correctness&quot;.
            &nbsp; The first requirement is that the data be &quot;well formed&quot; meaning that tags
            are syntactically correct and nested tags comply with a basic hierarchical data format.
            &nbsp; A more stringent requirement is based on the use defined data type declarations.
            &nbsp; XML data that specifies and obeys a data type definition (DTD) is said to be
            &quot;valid&quot;.&nbsp; An XML parser that looks for and uses a DTD in parsing the data
            is said to be a &quot;validating parser&quot;.&nbsp; Unfortunately, many of the code
            samples currently on the net are neither well formed or valid, in part because validating
            parsers have only recently been widely available.</font></p>
          </blockquote>
        </blockquote>
        <h2><font color="#0000A0"><a name="Software">Software</a> distributions</font></h2>
        <p><font color="#0000A0">DTD (data type defintion files)</font><ul>
          <li><font color="#0000A0"><a href="http://www.ibc.wustl.edu/bioXML/dtd/bioml.dtd">bioml.dtd</a>
            - a slightly edited version of the Proteomics <a href="#bioml">BIOML</a> data type
            definition that supports validated parsing of sequence data.&nbsp; See <a
            href="http://www.ibc.wustl.edu/bioXML/dtd/insulin2.xml">insulin2.xml</a> for a sample of
            XML data.</font></li>
        </ul>
        <p><font color="#0000A0"><a href="http://www.cpan.org/modules/by-module/XML/">XML perl
        modules</a> - The CPAN distribution of PERL modules for parsing and manipulating XML.
        &nbsp; Includes: </font><ul>
          <li><font color="#0000A0">XML::CGI - a module to convert CGI parameters to and from XML</font></li>
          <li><font color="#0000A0">XML::DOM - a Perl extension to XML::Parser. It adds a new 'Style'
            to XML::Parser,called 'Dom', that allows XML::Parser to build an Object Oriented data
            structure with a DOM Level 1 compliant interface.</font></li>
          <li><font color="#0000A0">XML::Dumper - a simple package to experiment with converting Perl
            data structures to XML and converting XML to perl data structures. </font></li>
          <li><font color="#0000A0">XML::Encoding - a subclass of XML::Parser, parses encoding map XML
            files. </font></li>
          <li><font color="#0000A0">XML::Generator is an extremely simple module to help in the
            generation of XML.</font></li>
          <li><font color="#0000A0">XML::Grove - provides simple objects for parsed XML documents. The
            objects may be modified but no checking is performed.</font></li>
          <li><font color="#0000A0"><a name="XML::Parser">XML::Parser</a> - a Perl extension interface
            to James Clark's XML parser, expat</font></li>
          <li><font color="#0000A0">XML::QL - an early implementation of a note published by the W3C
            called &quot;XML-QL: A Query Language for XML&quot;.</font></li>
          <li><font color="#0000A0">XML::XQL - a Perl extension that allows you to perform XQL queries
            on XML object trees. </font></li>
        </ul>
        <p><font color="#0000A0">Non-CPAN sites </font><ul>
          <li><font color="#0000A0"><a href="http://http://bio.perl.org/Core/Latest/">BioPERL</a>
            distribution</font></li>
          <li><font color="#0000A0"><a href="http://www.fastnetltd.ndirect.co.uk/Perl/">http://www.fastnetltd.ndirect.co.uk/Perl/</a>
            - XML::DBI and DBI::XML modules</font></li>
          <li><font color="#0000A0"><a href="http://www.xmlsoftware.com/">XMLsoftware.com</a> - site
            offers a mix of public domain, trial offer, and licensed software distributions for
            working with XML.</font></li>
          <li><a href="http://www.stud.ifi.uio.no/~larsga/linker/XMLtools.html"><font color="#0000A0">Free
            XML software distributions</font></a></li>
        </ul>
        <p><font color="#0000A0">Many of the modules are under active development so features and
        capabilities may evolve with time.</font></p>
        <h2><font color="#0000A0">Mailing lists and <a name="discussion">discussion</a> groups</font></h2>
        <ul>
          <li><font color="#0000A0"><a href="http://www.bioperl.org/MailList.html">BIO.PERL developers
            forum</a></font></li>
          <li><font color="#0000A0"><a href="http://www.xml.com/xml/pub/Guide/Mailing_Lists">XML.com's
            list of lists</a></font></li>
        </ul>
        <h2><font color="#0000A0"><a name="Related">Related</a> web sites</font></h2>
        <blockquote>
          <h3><a href="http://http://www.ibc.wustl.edu/bioXML/"><font color="#0000A0">bioXML</font></a></h3>
          <ul>
            <li><font color="#0000A0">The bioXML site is at present a mirror of this site.&nbsp; In the
              future we will add site specific data and software links related to Washington University
              specific activities.</font></li>
          </ul>
          <h3><font color="#0000A0">W3C</font></h3>
          <ul>
            <li><font color="#0000A0"><a href="http://www.ucc.ie/xml/">FAQ</a> - Frequently asked
              questions about XML, maintainted by the W3C XML special interest group</font></li>
            <li><font color="#0000A0"><a href="http://w3c.org/XML/">XML</a> - Central node for the
              development and dissemination of standards and information about XML.</font></li>
            <li><font color="#0000A0"><a href="http://w3c.org/RDF/">RDF</a> - Central node for
              development of web based metadata standards.</font></li>
          </ul>
          <h3><font color="#0000A0">XML tutorials and links</font></h3>
          <ul>
            <li><font color="#0000A0"><a href="http://www.xml.com/">XML.com</a> - a webzine devoted to
              XML.</font></li>
            <li><font color="#0000A0"><a href="http://www.oasis-open.org/cover/xml.html">Oasis</a> -
              Another central XML site with annotated specifications and links to conferences,
              tutorials, and related documentation.</font></li>
            <li><font color="#0000A0"><a href="http://www.webreference.com/authoring/languages/xml/">Webreference.com</a>
              XML links</font></li>
            <li><font color="#0000A0"><a
              href="http://www.microsoft.com/workshop/c-frame.htm#/xml/default.asp">Microsoft</a> XML
              tutorial (contains some nonstandard or not yet standardized material)</font></li>
            <li><font color="#0000A0"><a href="http://www.microsoft.com/xml">Microsoft</a> XML links and
              information</font></li>
            <li><font color="#0000A0"><a href="http://developer.netscape.com/tech/metadata/index.html">Netscape</a>
              XML links and information</font></li>
            <li><font color="#0000A0"><a
              href="http://www.netheaven.com/~coopercc/xmlparser/article.html">Clark Cooper </a>article
              on an XML::Parser</font></li>
            <li><font color="#0000A0"><a href="http://www.finetuning.com/parse.html">Finetuning.com</a>
              parser central page and links</font></li>
          </ul>
          <h3><font color="#0000A0">Bioinformatics</font></h3>
          <blockquote>
            <p><a href="http://visualgenomics.com/sbir/rfc.htm" name="bsml"><font color="#0000A0">BSML</font></a><font
            color="#0000A0"> - Bioinformatic Sequence Markup Language - a proposed data type
            definition for the representation of molecular biological data in XML developed by Visual
            Genomics and used as the basis for the BSML web browser developed and distributed by the
            same company.</font></p>
            <p><a href="http://www.proteometrics.com/BIOML/" name="bioml"><font color="#0000A0">BIOML</font></a><font
            color="#0000A0"> - Biopolymer Markup Language - a proposed data type definition for
            molecular biological sequence and structure data in XML developed by ProteoMetrics, LLC.</font></p>
            <p><font color="#0000A0"><a href="http://www.bioperl.org">BIO.PERL.ORG</a> - a site
            dedicated to the development of PERL tools for molecular biology computing, including
            multiple discussion forums, web archives, and pointers to software distributions</font></p>
            <p><font color="#0000A0"><a href="http://www.ibc.wustl.edu/service/standard.html">Standards
            and definitions</a> for bioinformatics computing</font></p>
          </blockquote>
        </blockquote>
        <blockquote>
          <h3><font color="#0000A0">PERL</font></h3>
          <blockquote>
            <p><font color="#0000A0"><a href="http://www.cpan.org/CPAN.html">CPAN</a> - the
            Comprehensive Perl Archive Network - a central distribution point for PERL modules on
            everything including XML parsing and database interfaces.</font></p>
          </blockquote>
        </blockquote>
        <h2><font color="#0000A0"><a name="Site">Site</a> maintenance</font></h2>
        <p><font color="#0000A0">This page is coordinated by <a
        href="http://www.ibc.wustl.edu/~states">David States</a> and mirrors content from the <a
        href="http://www.ibc.wustl.edu/bioXML">bioXML</a> site at the Institute for Biomedical
        Computing, Washington University in St. Louis. &nbsp; Comments and contributions welcome.</font></p>
        <p align="center"><font color="#0000A0"><em>Last updated <!--webbot bot="Timestamp"
        startspan s-type="EDITED" s-format="%B %d, %Y" -->January 07, 1999<!--webbot bot="Timestamp"
        i-checksum="32826" endspan --></em></font></td>
      </tr>
      <tr>
        <td width="234"><font face="Arial, Helvetica, sans-serif" size="+1"><a href="../">&lt;&lt;back</a>
        </font></td>
      </tr>
<!spacer >
      <tr>
        <td width="234"><img SRC="/images/transparent.gif" WIDTH="230" HEIGHT="1"></td>
      </tr>
    </table>
    </td>
<! end first main column>
<! vertical spacer bar>
    <td WIDTH="3" BGCOLOR="#000080"><img SRC="/images/transparent.gif" WIDTH="1" HEIGHT="1"></td>
  </tr>
</table>
<! ### START CVS DATESTAMP >

<p align="right">Last Modified: $Date$</p>
<! ### STOP CVS DATESTAMP >
<! ### START CVS REVISION >
<!-- $Id$ -->
<! ### STOP CVS REVISION >
</body>
</html>
